File: backend/src/application/use-cases/ResolveEmergency.ts
import { Emergency } from '../../domain/entities/Emergency';
import { IEmergencyRepository } from '../../domain/repositories/IEmergencyRepository';

export interface ResolveEmergencyDTO {
  emergencyId: string;
  resolvedBy: string;
}

export class ResolveEmergency {
  constructor(private readonly emergencyRepository: IEmergencyRepository) {}

  async execute(dto: ResolveEmergencyDTO): Promise<Emergency> {
    const emergency = await this.emergencyRepository.findById(dto.emergencyId);
    if (!emergency) {
      throw new Error('Emergency not found');
    }

    if (!emergency.isActive()) {
      throw new Error('Emergency is not active');
    }

    emergency.resolve(dto.resolvedBy);
    return this.emergencyRepository.update(emergency);
  }
}

File: backend-dotnet/src/Hestia.Application/UseCases/ResolveEmergency.cs
using Hestia.Domain.Entities;
using Hestia.Domain.Repositories;

namespace Hestia.Application.UseCases;

public class ResolveEmergency
{
    private readonly IEmergencyRepository _emergencyRepository;

    public ResolveEmergency(IEmergencyRepository emergencyRepository)
        => _emergencyRepository = emergencyRepository;

    public async Task<Emergency> ExecuteAsync(Guid emergencyId, Guid resolvedBy, CancellationToken ct = default)
    {
        var emergency = await _emergencyRepository.FindByIdAsync(emergencyId, ct)
            ?? throw new ApplicationException("Emergency not found");

        if (!emergency.IsActive())
            throw new ApplicationException("Emergency is not active");

        emergency.Resolve(resolvedBy);
        return await _emergencyRepository.UpdateAsync(emergency, ct);
    }
}
